File: src/main/java/fi/eonwe/wikilinks/WikiProcessor.java
package fi.eonwe.wikilinks;

import com.google.common.collect.Lists;
import fi.eonwe.wikilinks.fatpages.PagePointer;
import fi.eonwe.wikilinks.fatpages.WikiPage;
import fi.eonwe.wikilinks.fatpages.WikiPageData;
import fi.eonwe.wikilinks.fatpages.WikiRedirectPage;
import fi.eonwe.wikilinks.leanpages.BufferWikiPage;
import info.bliki.wiki.dump.IArticleFilter;
import info.bliki.wiki.dump.Siteinfo;
import info.bliki.wiki.dump.WikiArticle;
import info.bliki.wiki.dump.WikiPatternMatcher;
import info.bliki.wiki.dump.WikiXMLParser;
import net.openhft.koloboke.collect.map.hash.HashObjObjMap;
import net.openhft.koloboke.collect.map.hash.HashObjObjMaps;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Comparator;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;

/**
 */
public class WikiProcessor {

    public static List<BufferWikiPage> readPages(InputStream input) {
        WikiProcessor processor = new WikiProcessor();
        HashObjObjMap<String, PagePointer> pages = processor.preProcess(input);
//        printStatistics(pages);
        WikiProcessor.dropRedirectLoops(pages);
//        printStatistics(pages);
        List<BufferWikiPage> packedPages = WikiProcessor.packPages(pages);
        return packedPages;
    }

    public HashObjObjMap<String, PagePointer> preProcess(InputStream input) {
        final HashObjObjMap<String, PagePointer> titleToPage = HashObjObjMaps.newMutableMap(12_000_000);
        try {
            WikiXMLParser parser = new WikiXMLParser(input, new IArticleFilter() {
                @Override
                public void process(WikiArticle article, Siteinfo siteinfo) {
                    if (article.isMain()) {
                        String text = article.getText();
                        if (text == null) text = "";
                        WikiPatternMatcher matcher = new WikiPatternMatcher(text);
                        int id = Integer.parseInt(article.getId());
                        WikiPage page;
                        if (matcher.isRedirect()) {
                            page = new WikiRedirectPage(article.getTitle().intern(), id, matcher.getRedirectText().intern());
                            fixPagePointers(titleToPage, page);
                        } else {
                            String[] links = matcher.getLinks().stream().filter(l -> !l.isEmpty()).map(WikiProcessor::possiblyCapitalize).distinct().toArray(String[]::new);
                            PagePointer[] pointerLinks = new PagePointer[links.length];
                            for (int i = 0; i < links.length; i++) {
                                String link = links[i];
                                PagePointer ptr = titleToPage.get(link);
                                if (ptr == null) {
                                    ptr = new PagePointer(null);
                                    titleToPage.put(link.intern(), ptr);
                                }
                                pointerLinks[i] = ptr;
                            }
                            page = new WikiPageData(article.getTitle().intern(), id, pointerLinks);
                            fixPagePointers(titleToPage, page);
                        }
                    }
                }
            });
            parser.parse();
            return titleToPage;
        } catch (SAXException | IOException e) {
            return titleToPage;
        }
    }

    private static String possiblyCapitalize(String linkName) {
        if (linkName.length() != 0 && !Character.isUpperCase(linkName.charAt(0))) {
            char[] chars = linkName.toCharArray();
            chars[0] = Character.toUpperCase(chars[0]);
            return new String(chars);
        }
        return linkName;
    }

    private static void fixPagePointers(HashObjObjMap<String, PagePointer> titleToPage, WikiPage page) {
        PagePointer pointer = titleToPage.get(page.getTitle());
        if (pointer != null) {
            pointer.page = page;
        } else {
            pointer = new PagePointer(page);
            titleToPage.put(page.getTitle(), pointer);
        }
    }

    public static void dropRedirectLoops(HashObjObjMap<String, PagePointer> map) {
        map.values().stream().filter(p -> p.page != null && p.page.isRedirect()).forEach(p -> p.page = endSomewhere(p, map, null) ? p.page : null);
    }

    private static boolean endSomewhere(PagePointer redirect, HashObjObjMap<String, PagePointer> map, IdentityHashMap<WikiPage, Boolean> visited) {
        WikiPage immediateTarget = redirect.page;
        if (immediateTarget instanceof WikiRedirectPage) {
            if (visited == null) {
                visited = new IdentityHashMap<>();
            }
            if (visited.containsKey(immediateTarget)) {
                return false;
            } else {
                visited.put(immediateTarget, Boolean.TRUE);
            }
            WikiRedirectPage redirectPage = (WikiRedirectPage) immediateTarget;
            PagePointer redirectPointer = map.get(redirectPage.getTarget());
            if (redirectPointer == null) {
                return false;
            } else {
                return endSomewhere(redirectPointer, map, visited);
            }
        } else {
            return true;
        }
    }

    public static void printStatistics(Map<String, PagePointer> map) {
        int articleCount = 0;
        int redirectCount = 0;
        int linkCount = 0;
        int nullLinkCount = 0;
        for (PagePointer ptr : map.values()) {
            WikiPage page = ptr.page;
            if (page == null) {
                nullLinkCount++;
                continue;
            }
            if (page instanceof WikiRedirectPage) {
                redirectCount++;
            } else if (page instanceof WikiPageData) {
                articleCount++;
                WikiPageData p = (WikiPageData) page;
                for (PagePointer linkPointer : p.getLinks()) {
                    WikiPage linkedPage = linkPointer.page;
                    if (linkedPage == null) {
                        nullLinkCount++;
                    } else {
                        linkCount++;
                    }
                }
            }
        }
        System.out.printf(
                "Articles: %d%n" +
                "Redirects: %d%n" +
                "Links: %d (/ article: %.2f)%n" +
                "Null links %d%n",
                articleCount, redirectCount,
                linkCount, linkCount / (double) articleCount,
                nullLinkCount);
    }

    private static final int[] EMPTY_ARRAY = new int[0];

    public static List<BufferWikiPage> packPages(HashObjObjMap<String, PagePointer> map) {
        List<BufferWikiPage> list = Lists.newArrayListWithCapacity(map.size());
        map.forEach((title, ptr) -> {
            WikiPage page = ptr.page;
            if (page != null) {
                int[] links;
                boolean isRedirect;
                if (page instanceof WikiRedirectPage) {
                    WikiRedirectPage redirectPage = (WikiRedirectPage) page;
                    String target = redirectPage.getTarget();
                    PagePointer pagePointer = map.get(target);
                    if (pagePointer == null || pagePointer.page == null) {
                        links = EMPTY_ARRAY;
                    } else {
                        links = new int[] { pagePointer.page.getId() };
                    }
                    isRedirect = true;
                } else {
                    WikiPageData pageData = (WikiPageData) page;
                    links = Arrays.stream(pageData.getLinks()).filter(p -> p.page != null).mapToInt(p -> p.page.getId()).distinct().toArray();
                    if (links.length == 0) links = EMPTY_ARRAY;
                    Arrays.sort(links);
                    isRedirect = false;
                }
                BufferWikiPage packedPage = BufferWikiPage.createFrom(page.getId(), links, title, isRedirect);
                list.add(packedPage);
            }
        });
        list.sort(Comparator.comparingLong(BufferWikiPage::getId));
        return list;
    }
}

File: src/main/kotlin/fi/eonwe/wikilinks/WikiProcessor.kt
package fi.eonwe.wikilinks

import fi.eonwe.wikilinks.fatpages.PagePointer
import fi.eonwe.wikilinks.fatpages.WikiPage
import fi.eonwe.wikilinks.fatpages.WikiPageData
import fi.eonwe.wikilinks.fatpages.WikiRedirectPage
import fi.eonwe.wikilinks.leanpages.BufferWikiPage
import info.bliki.wiki.dump.WikiPatternMatcher
import info.bliki.wiki.dump.WikiXMLParser
import net.openhft.koloboke.collect.map.hash.HashObjObjMap
import net.openhft.koloboke.collect.map.hash.HashObjObjMaps
import java.io.InputStream
import java.util.Arrays
import java.util.IdentityHashMap
import java.util.concurrent.atomic.AtomicBoolean

/**
 */
class WikiProcessor private constructor() {
    data class GraphStatistics(
        val articleCount: Int,
        val redirectCount: Int,
        val linkCount: Int,
        val nullLinkCount: Int
    )

    data class ReadPagesResult(
        val pages: MutableList<BufferWikiPage>,
        val beforeRedirectCleanup: GraphStatistics,
        val afterRedirectCleanup: GraphStatistics
    )

    fun preProcess(input: InputStream): HashObjObjMap<String, PagePointer> {
        val titleToPage = HashObjObjMaps.newMutableMap<String, PagePointer>(12000000)
        var nextInternalId = 0

        fun nextPageId(): Int {
            if (nextInternalId == Int.MAX_VALUE) {
                throw IllegalStateException("Too many pages to index with 32-bit ids")
            }
            return nextInternalId++
        }

        val parser = WikiXMLParser(input) { article, siteinfo ->
            if (article.isMain) {
                val text = article.text ?: ""
                val matcher = WikiPatternMatcher(text)
                // Wikimedia page ids can be larger than Int. We use a compact internal id instead.
                val id = nextPageId()
                if (matcher.isRedirect) {
                    val page = WikiRedirectPage(article.title, id, matcher.redirectText)
                    fixPagePointers(titleToPage, page)
                } else {
                    val links = matcher.links.asSequence()
                        .map { linkName -> possiblyCapitalize(linkName) }
                        .distinct()
                    val pointerLinks = buildList {
                        links.forEach { link ->
                            var ptr = titleToPage[link]
                            if (ptr == null) {
                                ptr = PagePointer(null)
                                titleToPage[link] = ptr
                            }
                            add(ptr)
                        }
                    }
                    val page = WikiPageData(article.title, id, pointerLinks)
                    fixPagePointers(titleToPage, page)
                }
            }
        }
        parser.parse()
        return titleToPage
    }

    companion object {
        private val xmlLimitsConfigured = AtomicBoolean(false)

        private fun configureXmlParserLimitsForTrustedWikiDump() {
            if (!xmlLimitsConfigured.compareAndSet(false, true)) {
                return
            }
            // Allow very large trusted Wikimedia dumps to parse without JAXP default caps.
            setSystemPropertyIfMissing("jdk.xml.totalEntitySizeLimit", "0")
            setSystemPropertyIfMissing("jdk.xml.entityExpansionLimit", "0")
            setSystemPropertyIfMissing("jdk.xml.maxGeneralEntitySizeLimit", "0")
            setSystemPropertyIfMissing("jdk.xml.maxParameterEntitySizeLimit", "0")
        }

        private fun setSystemPropertyIfMissing(name: String, value: String) {
            if (System.getProperty(name).isNullOrBlank()) {
                System.setProperty(name, value)
            }
        }

        fun readPages(input: InputStream): MutableList<BufferWikiPage> {
            return readPagesWithStats(input).pages
        }

        fun readPagesWithStats(input: InputStream): ReadPagesResult {
            configureXmlParserLimitsForTrustedWikiDump()
            val processor = WikiProcessor()
            val pages = processor.preProcess(input)
            val before = gatherStatistics(pages)
            dropRedirectLoops(pages)
            val after = gatherStatistics(pages)
            val packedPages = packPages(pages)
            return ReadPagesResult(
                pages = packedPages,
                beforeRedirectCleanup = before,
                afterRedirectCleanup = after
            )
        }

        private fun possiblyCapitalize(linkName: String): String {
            if (linkName.isNotEmpty() && !Character.isUpperCase(linkName[0])) {
                val chars = linkName.toCharArray()
                chars[0] = chars[0].uppercaseChar()
                return String(chars)
            }
            return linkName
        }

        fun fixPagePointers(titleToPage: HashObjObjMap<String, PagePointer>, page: WikiPage) {
            var pointer = titleToPage[page.title()]
            if (pointer != null) {
                pointer.page = page
            } else {
                pointer = PagePointer(page)
                titleToPage.put(page.title(), pointer)
            }
        }

        fun dropRedirectLoops(map: HashObjObjMap<String, PagePointer>) {
            map.values.asSequence()
                .filter { p ->
                    val page = p.page
                    page != null && page.isRedirect
                }.forEach { p ->
                    val didFindEndPage = endSomewhere(p, map, null)
                    if (!didFindEndPage) {
                        p.page = null
                    }
                }
        }

        private fun endSomewhere(
            redirect: PagePointer,
            map: HashObjObjMap<String, PagePointer>,
            visited: IdentityHashMap<WikiPage, Boolean>?
        ): Boolean {
            var visited = visited
            val immediateTarget = redirect.page
            if (immediateTarget is WikiRedirectPage) {
                if (visited == null) {
                    visited = IdentityHashMap<WikiPage, Boolean>()
                }
                if (visited.containsKey(immediateTarget)) {
                    return false
                } else {
                    visited.put(immediateTarget, true)
                }
                val redirectPage = immediateTarget
                val redirectPointer = map[redirectPage.target]
                if (redirectPointer == null) {
                    return false
                } else {
                    return endSomewhere(redirectPointer, map, visited)
                }
            } else {
                return true
            }
        }

        fun gatherStatistics(map: MutableMap<String?, PagePointer>): GraphStatistics {
            var articleCount = 0
            var redirectCount = 0
            var linkCount = 0
            var nullLinkCount = 0
            for (ptr in map.values) {
                val page = ptr.page
                if (page == null) {
                    nullLinkCount++
                    continue
                }
                if (page is WikiRedirectPage) {
                    redirectCount++
                } else if (page is WikiPageData) {
                    articleCount++
                    val p = page
                    for (linkPointer in p.links) {
                        val linkedPage = linkPointer.page
                        if (linkedPage == null) {
                            nullLinkCount++
                        } else {
                            linkCount++
                        }
                    }
                }
            }
            return GraphStatistics(
                articleCount = articleCount,
                redirectCount = redirectCount,
                linkCount = linkCount,
                nullLinkCount = nullLinkCount
            )
        }

        fun printStatistics(map: MutableMap<String?, PagePointer>) {
            printStatistics(gatherStatistics(map))
        }

        fun printStatistics(stats: GraphStatistics) {
            System.out.printf(
                "Articles: %d%n" +
                        "Redirects: %d%n" +
                        "Links: %d (/ article: %.2f)%n" +
                        "Null links %d%n",
                stats.articleCount, stats.redirectCount,
                stats.linkCount, stats.linkCount / stats.articleCount.toDouble(),
                stats.nullLinkCount
            )
        }

        private val EMPTY_ARRAY = IntArray(0)

        fun packPages(map: HashObjObjMap<String, PagePointer>): MutableList<BufferWikiPage> {
            val list = mutableListOf<BufferWikiPage>()
            map.forEach { (title: String, ptr: PagePointer) ->
                val page = ptr.page
                if (page != null) {
                    var links: IntArray
                    val isRedirect = if (page is WikiRedirectPage) {
                        val redirectPage = page
                        val target = redirectPage.target
                        val pagePointer = map[target]
                        if (pagePointer == null || pagePointer.page == null) {
                            links = EMPTY_ARRAY
                        } else {
                            links = intArrayOf(pagePointer.page!!.id())
                        }
                        true
                    } else {
                        val pageData = page as WikiPageData
                        links = pageData.links
                            .stream()
                            .map { it.page }
                            .filter { it != null }
                            .mapToInt { p -> p!!.id() }
                            .distinct()
                            .toArray()
                        if (links.isEmpty()) links = EMPTY_ARRAY
                        Arrays.sort(links)
                        false
                    }
                    val packedPage = BufferWikiPage.createFrom(page.id(), links, title, isRedirect)
                    list.add(packedPage)
                }
            }
            list.sortBy { it.id }
            return list
        }
    }
}
